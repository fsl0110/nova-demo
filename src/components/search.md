### A Search-Component

#### With

```jsx
<Search isLoading={false} fetchData={() => null} />
```

```jsx
<Search isLoading={true} fetchData={() => null} />
```
