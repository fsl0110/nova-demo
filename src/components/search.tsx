import React from "react";
import classNames from "classnames";
import dompurify from "dompurify";
import { SharedProps } from "../typings";

const sanitizer = dompurify.sanitize;

export interface Props extends SharedProps {
  /** Is loading data */
  isLoading: boolean;
  /** ClickHandler for fetching data */
  fetchData: (term: string) => void;
}

export class Search extends React.Component<Props> {
  private searchInputRef = React.createRef<HTMLInputElement>();

  handleSubmitClick = () => {
    this.props.fetchData(sanitizer(this.searchInputRef.current!.value));
  };

  handleEnterClick = (e: any) => {
    if (e.keyCode === 13 && !this.props.isLoading) {
      this.props.fetchData(sanitizer(this.searchInputRef.current!.value));
    }
  };

  render() {
    const { isLoading, className } = this.props;

    return (
      <div className={classNames(className, "search")}>
        <input
          className="search__input"
          type="text"
          aria-label="Searchfield for fiiltering food products"
          ref={this.searchInputRef}
          onKeyDown={this.handleEnterClick}
          placeholder="e.g. ice cream"
        />
        <button
          className="search__button"
          aria-label="Searchbutton for filtering food products"
          onClick={this.handleSubmitClick}
          disabled={isLoading}
        >
          <span>&#9906;</span>
        </button>
      </div>
    );
  }
}

export default Search;
