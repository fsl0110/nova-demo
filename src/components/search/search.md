### Search for filtering data

#### Search default

```jsx
<Search isLoading={false} fetchData={() => null} />
```

#### Search while is loading

```jsx
<Search isLoading={true} fetchData={() => null} />
```
