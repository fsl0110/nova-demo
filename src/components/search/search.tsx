import React, {
  createRef,
  PureComponent,
  HTMLAttributes,
  KeyboardEvent
} from "react";
import classNames from "classnames";
import dompurify from "dompurify";
import { Button } from "..";

const sanitizer = dompurify.sanitize;

export interface Props extends HTMLAttributes<HTMLDivElement> {
  /** Is loading data */
  isLoading?: boolean;
  /** ClickHandler for fetching data */
  fetchData: (term?: string) => void;
}

export class Search extends PureComponent<Props> {
  private searchInputRef = createRef<HTMLInputElement>();

  handleEnterClick = (e: KeyboardEvent<HTMLInputElement>) => {
    if (e.keyCode === 13 && !this.props.isLoading) {
      this.handleFetch();
    }
  };

  handleFetch = () => {
    const value =
      this.searchInputRef.current && this.searchInputRef.current.value
        ? this.searchInputRef.current.value
        : ""; // this is a quickfix to avoid null error in test
    this.props.fetchData(sanitizer(value));
  };

  render() {
    const { isLoading, className } = this.props;

    return (
      <div className={classNames(className, "search")} data-test="Search">
        <input
          aria-label="Searchfield for filtering food products"
          className="search__input"
          type="text"
          ref={this.searchInputRef}
          onKeyDown={this.handleEnterClick}
          placeholder="e.g. ice cream"
        />
        <Button
          aria-label="Searchbutton for filtering food products"
          className="search__button"
          iconName="search"
          iconSize="l"
          isLoading={isLoading}
          onClick={this.handleFetch}
          data-test="search-button"
        />
      </div>
    );
  }
}

export default Search;
