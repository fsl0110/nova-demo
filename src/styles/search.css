.search {
  display: flex;
  justify-content: space-between;
  align-items: center;
  height: 40px;
  border: 2px solid rgb(32, 32, 32);
  border-radius: 22px;
  width: 300px;
  margin: 0 auto 2rem auto;
}

.search__input {
  padding: 0 1rem;
  border: none;
  width: 100%;
  font-size: 1.1rem;
  border-top-left-radius: 20px;
  border-bottom-left-radius: 20px;
  caret-color: black;
}

.search:focus-within {
  border-color: pink;
  border-width: 2px;
}

.search__button {
  border: none;
  background-color: rgba(0, 0, 0, 0);
  width: 60px;
  height: 50px;
  font-size: 2.1rem;
  transform: rotate(45deg) translateX(7px) translatey(-7px);
  border: 1px dotted transparent;
}

.search__button:focus {
  border: 1px dotted gray;
  border-radius: 4px;
}
.search__button:disabled {
  cursor: default;
}
