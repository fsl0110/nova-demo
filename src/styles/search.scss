.search {
  display: flex;
  justify-content: space-between;
  align-items: center;
  height: 32px;
  border: 2px solid rgb(32, 32, 32);
  border-radius: 22px;
  width: 200px;

  margin-bottom: 0.5rem;
}

.search__input {
  padding: 0 1rem;
  border: none;
  width: 100%;
  font-size: 1.1rem;
  border-top-left-radius: 20px;
  border-bottom-left-radius: 20px;
  caret-color: black;
}

.search:focus-within {
  border-color: pink;
  border-width: 2px;
}

.search__button {
  border: 1px dotted transparent;
  height: 40px;
  padding-right: 3px;
}

.search__button:focus {
  border: 1px dashed rgb(32, 32, 32);
  border-radius: 50%;
}

.search__button:disabled {
  cursor: progress;
}
