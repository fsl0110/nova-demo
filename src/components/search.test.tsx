import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { Search, Props } from "./search";

const testProps: Props = {
  isLoading: false,
  fetchData: jest.fn()
};

const testCases: Array<[string, Props]> = [
  [
    "renders with isLoading=true",
    {
      isLoading: true,
      fetchData: jest.fn()
    }
  ],
  [
    "renders with isLoading=false",
    {
      isLoading: false,
      fetchData: jest.fn()
    }
  ]
];

describe("<Search>", () => {
  it("renders without crashing", () => {
    shallow(<Search {...testProps} />);
  });

  testCases.forEach(([item, options]) => {
    it(item, () => {
      const mock = shallow(<Search {...options} />);
      expect(toJson(mock)).toMatchSnapshot();
    });
  });
});
