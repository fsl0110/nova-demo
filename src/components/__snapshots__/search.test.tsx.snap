// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search> renders with isLoading=false 1`] = `
<div
  className="search"
>
  <input
    aria-label="Searchfield for fiiltering food products"
    className="search__input"
    onKeyDown={[Function]}
    placeholder="e.g. ice cream"
    type="text"
  />
  <button
    aria-label="Searchbutton for filtering food products"
    className="search__button"
    disabled={false}
    onClick={[Function]}
  >
    <span>
      ⚲
    </span>
  </button>
</div>
`;

exports[`<Search> renders with isLoading=true 1`] = `
<div
  className="search"
>
  <input
    aria-label="Searchfield for fiiltering food products"
    className="search__input"
    onKeyDown={[Function]}
    placeholder="e.g. ice cream"
    type="text"
  />
  <button
    aria-label="Searchbutton for filtering food products"
    className="search__button"
    disabled={true}
    onClick={[Function]}
  >
    <span>
      ⚲
    </span>
  </button>
</div>
`;
